CC Macro name conflict when two packages have the same name

580ea82b4eda211c252bd60f4a5f4333658bc5c0 is a unit test reproducing the issue

# Problem statment

The haskell package name of an `haskell_library` is based on the bazel rule name, so `//tests:bytestring` and `//tests/cpp_macro_conflict:bytestring` have the same name `bytestring`.

When enabling haskell extension `CPP`, ghc is exporting macros `VERSION_pkgname` (e.g. `VERSION_bytestring`), this results in macro name conflicts when a `haskell_library` or `haskell_binary` have two `deps` with the same bazel name.

Using code from 580ea82b4eda211c252bd60f4a5f4333658bc5c0, I got this error when building with `bazel build //tests/cpp_macro_conflict/...`

```
/run/user/1000/ghc8953_0/ghc_2.h:13:0: error:
     error: "VERSION_bytestring" redefined [-Werror]
     #define VERSION_bytestring ""
     
   |
13 | #define VERSION_bytestring ""
   | ^

/run/user/1000/ghc8953_0/ghc_2.h:7:0: error:
     note: this is the location of the previous definition
     #define VERSION_bytestring "0.10.8.2"
     
  |
7 | #define VERSION_bytestring "0.10.8.2"
  | ^

/run/user/1000/ghc8953_0/ghc_2.h:14:0: error:
     error: "MIN_VERSION_bytestring" redefined [-Werror]
     #define MIN_VERSION_bytestring(major1,major2,minor) (\
     
   |
14 | #define MIN_VERSION_bytestring(major1,major2,minor) (\
   | ^

/run/user/1000/ghc8953_0/ghc_2.h:8:0: error:
     note: this is the location of the previous definition
     #define MIN_VERSION_bytestring(major1,major2,minor) (\
     
  |
8 | #define MIN_VERSION_bytestring(major1,major2,minor) (\
  | ^
```

Note: this is an error due to `-Warn`, but usually that's just annoying warnings.

One *simple* workaround is to tell user to avoid package name conflict when using `CPP`.

# Discussion

Version name is useful for library from the outside world (i.e. hackage) because they may implement different behaviors depending on some version of their dependencies, which are also coming from the outside world. These external libraries can still be built inside our bazel repository (i.e. using Hazel). These external libraries depends on these macro and on the correct package name. For them we need to export the version macro and keep the official package name. External libraries have a version number (and hazel exposes it using `version` attribute).

Libraries defined in a bazel repository may not be concerned by version number, that the point of  a mega repo, you control all your dependencies and everything must be coherent inside the mega repo. Hence users won't need these macros for local libraries and we can disable them.

The undocumented option `-fno-macro-version` (https://ghc.haskell.org/trac/ghc/ticket/11763) works for all the macros during one build operation. This does not allow a per-package choice. For example, if a rule depends on both the "hackage" `bytestring` and a local `bytestring`, we cannot enable macro generation for the former and not for the later.

## Design proposal

Each package controls if its macro version will be exported when the package is used as a dependency. Using for example an `export_macros` attribute. The exported macro depends on the `version` attribute.

Users will still have to change their naming scheme but only if they want to export the version macro of two packages with the same name.

For example:

```python
# This is the hackage "bytestring" library
haskell_library(
   name = "bytestring",
   version = "1.2.3",
   export_macros = True, # can be deduced from the presence of `version`
)

# This is a local library
haskell_library(
   name = "foo",
   export_macros = False, # can be deduced from the absence of `version` 
)

haskell_library(
   name = "bar",
   deps = [":bytestring", ":foo"],
)
```

Here, `bar` will be built with version macro for `bytestring` but not for `foo`.

### Implementation details

We'll have to use `-fno-macro-version` for all build and then manually export the needed macros. 

We'll have to manually replicate the macro export mechanisms, AFAIK there is no flag to tell GHC to export the version macro per package.

## Alternative design 0

We can tell user that, if they want to use CPP macro, they need to avoid package name conflict. They will need to change their naming scheme, but there is no work from our side (except documentation).

Simple, but forces the user to change its naming scheme.

## Design 1: disable all macros on a rule

This is the solution with the simplest implementation, using `-fno-macro-version`, we can disable all the macro generation during the rule execution. For example:

```python
# This is the hackage "bytestring" library, it needs the version macro to build
haskell_library(
   name = "bytestring",
   deps = [":text", ":lens"], # others dependencies from hackage
   enable_version_macro = True,
)

# This is a local library which does not need versino macro to build
haskell_library(
   name = "foo",
   enable_version_macro = False
)
```

Here `bytestring` will be built with the macro version for `lens` and `text`. However `foo` will be built without any macro version.

I don't like this approach, because disabling all the macro for a build changes nothing compared to the current status of rules_haskell, except that it will disable the warnings.

--- a/haskell/haskell.bzl
+++ b/haskell/haskell.bzl
@@ -119,6 +119,9 @@ def _mk_binary_rule(**kwargs):
                 allow_single_file = FileType([".hs", ".hsc", ".lhs"]),
                 doc = "File containing `Main` module (deprecated).",
             ),
+            version = attr.string(
+                doc = "Executable version. If this is specified, CPP version macros will be generated for this build.",
+            ),
             _dummy_static_lib = attr.label(
                 default = Label("@io_tweag_rules_haskell//haskell:dummy_static_lib"),
                 allow_single_file = True,
@@ -189,7 +192,7 @@ haskell_library = rule(
         ),
         version = attr.string(
             doc = """Library version. Not normally necessary unless to build a library
-            originally defined as a Cabal package.""",
+            originally defined as a Cabal package. If this is specified, CPP version macro will be generated.""",
         ),
     ),
     outputs = {

--- a/haskell/lint.bzl
+++ b/haskell/lint.bzl
@@ -59,6 +59,7 @@ def _haskell_lint_aspect_impl(target, ctx):
         use_direct = False,
         use_my_pkg_id = None,
         custom_package_caches = None,
+        version = ctx.rule.attr.version,
     ))
 
     sources = set.to_list(

--- a/haskell/private/actions/compile.bzl
+++ b/haskell/private/actions/compile.bzl
@@ -65,7 +65,7 @@ def _process_hsc_file(hs, cc, hsc_file):
 
     return hs_out, idir
 
-def _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_srcs, compiler_flags, with_profiling, my_pkg_id = None):
+def _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_srcs, compiler_flags, with_profiling, my_pkg_id, version):
     """Declare default compilation targets and create default compiler arguments.
 
     Returns:
@@ -129,6 +129,7 @@ def _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_sr
         use_direct = True,
         use_my_pkg_id = my_pkg_id,
         custom_package_caches = None,
+        version = version,
     ))
 
     header_files = []
@@ -269,7 +270,7 @@ def _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_sr
         ),
     )
 
-def compile_binary(hs, cc, java, dep_info, srcs, ls_modules, import_dir_map, extra_srcs, compiler_flags, with_profiling, main_function):
+def compile_binary(hs, cc, java, dep_info, srcs, ls_modules, import_dir_map, extra_srcs, compiler_flags, with_profiling, main_function, version):
     """Compile a Haskell target into object files suitable for linking.
 
     Returns:
@@ -279,7 +280,7 @@ def compile_binary(hs, cc, java, dep_info, srcs, ls_modules, import_dir_map, ext
         modules: set of module names
         source_files: set of Haskell source files
     """
-    c = _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_srcs, compiler_flags, with_profiling)
+    c = _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_srcs, compiler_flags, with_profiling, my_pkg_id = None, version = version)
     c.args.add(["-main-is", main_function])
 
     hs.toolchain.actions.run_ghc(
@@ -335,7 +336,7 @@ def compile_library(hs, cc, java, dep_info, srcs, ls_modules, other_modules, exp
         source_files: set of Haskell module files
         import_dirs: import directories that should make all modules visible (for GHCi)
     """
-    c = _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_srcs, compiler_flags, with_profiling, my_pkg_id = my_pkg_id)
+    c = _compilation_defaults(hs, cc, java, dep_info, srcs, import_dir_map, extra_srcs, compiler_flags, with_profiling, my_pkg_id = my_pkg_id, version = my_pkg_id.version)
 
     hs.toolchain.actions.run_ghc(
         hs,

--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -84,7 +84,8 @@ def link_binary(
         compiler_flags,
         objects_dir,
         with_profiling,
-        dummy_static_lib):
+        dummy_static_lib,
+        version):
     """Link Haskell binary from static object files.
 
     Returns:
@@ -138,6 +139,7 @@ def link_binary(
         use_direct = False,
         use_my_pkg_id = None,
         custom_package_caches = None,
+        version = version,
     ))
 
     _add_external_libraries(args, dep_info.external_libraries.values())
@@ -314,6 +316,7 @@ def link_library_dynamic(hs, cc, dep_info, extra_srcs, objects_dir, my_pkg_id):
         use_direct = False,
         use_my_pkg_id = None,
         custom_package_caches = None,
+        version = my_pkg_id.version if my_pkg_id else None,
     ))
 
     _add_external_libraries(args, dep_info.external_libraries.values())

--- a/haskell/private/actions/repl.bzl
+++ b/haskell/private/actions/repl.bzl
@@ -33,6 +33,7 @@ def build_haskell_repl(
         build_info,
         output,
         package_caches,
+        version,
         lib_info = None,
         bin_info = None):
     """Build REPL script.
@@ -58,6 +59,7 @@ def build_haskell_repl(
         use_direct = False,
         use_my_pkg_id = None,
         custom_package_caches = package_caches,
+        version = version,
     )
 
     if lib_info != None:

--- a/haskell/private/haskell_impl.bzl
+++ b/haskell/private/haskell_impl.bzl
@@ -77,6 +77,7 @@ use the 'haskell_import' rule instead.
         compiler_flags = ctx.attr.compiler_flags,
         with_profiling = False,
         main_function = ctx.attr.main_function,
+        version = ctx.attr.version,
     )
 
     c_p = None
@@ -100,6 +101,7 @@ use the 'haskell_import' rule instead.
             compiler_flags = ctx.attr.compiler_flags,
             with_profiling = True,
             main_function = ctx.attr.main_function,
+            version = ctx.attr.version,
         )
 
     binary = link_binary(
@@ -111,6 +113,7 @@ use the 'haskell_import' rule instead.
         c_p.objects_dir if with_profiling else c.objects_dir,
         with_profiling,
         ctx.file._dummy_static_lib,
+        version = ctx.attr.version,
     )
 
     solibs = set.union(
@@ -137,6 +140,7 @@ use the 'haskell_import' rule instead.
         repl_ghci_args = ctx.attr.repl_ghci_args,
         output = ctx.outputs.repl,
         package_caches = dep_info.package_caches,
+        version = ctx.attr.version,
         build_info = build_info,
         bin_info = bin_info,
     )
@@ -308,6 +312,7 @@ use the 'haskell_import' rule instead.
             compiler_flags = ctx.attr.compiler_flags,
             output = ctx.outputs.repl,
             package_caches = dep_info.package_caches,
+            version = ctx.attr.version,
             build_info = build_info,
             lib_info = lib_info,
         )

--- a/haskell/private/packages.bzl
+++ b/haskell/private/packages.bzl
@@ -2,12 +2,13 @@
 
 load(":private/set.bzl", "set")
 
-def expose_packages(build_info, lib_info, use_direct, use_my_pkg_id, custom_package_caches):
+def expose_packages(build_info, lib_info, use_direct, use_my_pkg_id, custom_package_caches, version):
     """
     Returns the list of command line argument which should be passed
     to GHC in order to enable haskell packages.
 
     build_info: is common to all builds
+    version: if the rule contains a version, we will export the CPP version macro
 
     All the other arguments are not understood well:
 
@@ -16,7 +17,6 @@ def expose_packages(build_info, lib_info, use_direct, use_my_pkg_id, custom_pack
     use_my_pkg_id: only used for one specific task in compile.bzl
     custom_package_caches: override the package_caches of build_info, used only by the repl
     """
-
     args = [
         # In compile.bzl, we pass this just before all -package-id
         # arguments. Not doing so leads to bizarre compile-time failures.
@@ -26,11 +26,21 @@ def expose_packages(build_info, lib_info, use_direct, use_my_pkg_id, custom_pack
         "-hide-all-packages",
     ]
 
-    # Expose all prebuilt dependencies
+    has_version = version != None and version != ""
 
+    if not has_version:
+        args.extend([
+            # Macro version are disabled for all packages by default
+            # and enabled for package with version
+            # see https://github.com/tweag/rules_haskell/issues/414
+            "-fno-version-macros",
+        ])
+
+    # Expose all prebuilt dependencies
+    #
     # We have to remember to specify all (transitive) wired-in
     # dependencies or we can't find objects for linking
-
+    #
     # XXX: This should be really dep_info.direct_prebuilt_deps, but since we
     # cannot add prebuilt_dependencies to the "depends" field on package
     # registration (see a comment there), we have to pass all transitive

--- a/tests/cpp_macro_conflict/BS.hs
+++ b/tests/cpp_macro_conflict/BS.hs
@@ -1,0 +1,1 @@
+module BS where

--- a/tests/cpp_macro_conflict/BUILD
+++ b/tests/cpp_macro_conflict/BUILD
@@ -1,0 +1,29 @@
+package(default_testonly = 1)
+
+load(
+    "@io_tweag_rules_haskell//haskell:haskell.bzl",
+    "haskell_library",
+    "haskell_binary",
+)
+
+# empty library with package name "bytestring"
+haskell_library(
+  name = "bytestring",
+  srcs = ["BS.hs"],
+  deps = ["//tests:base"],
+)
+
+# This depends on two packages "bytestring"
+# There should be no CPP macro conflict
+haskell_binary(
+  name = "macro_conflict",
+  srcs = ["Main.hs"],
+  deps = ["//tests:base", "//tests:bytestring", ":bytestring"],
+  compiler_flags = ["-XCPP", "-Werror"] + select({
+      # clang on darwin fails because of unused command line argument, it fails because of -Werror
+      "@bazel_tools//src/conditions:darwin": [
+          "-optP-Wno-unused-command-line-argument",
+      ],
+      "//conditions:default": [],
+  }),
+)

--- a/tests/cpp_macro_conflict/Main.hs
+++ b/tests/cpp_macro_conflict/Main.hs
@@ -1,0 +1,4 @@
+import qualified Data.ByteString
+import BS
+
+main = putStrLn "hello"

--- a/tests/package-name/BUILD
+++ b/tests/package-name/BUILD
@@ -20,6 +20,7 @@ haskell_test(
     name = "bin",
     size = "small",
     srcs = ["Main.hs"],
+    version = "0.0.0", # This flags triggers the `MIN_VERSION` macro generation
     deps = [
         ":lib-a_Z",
         "//tests:base",

